qutebrowser/browser/qtnetworkdownloads.py

```diff
@@ -0,0 +1,473 @@
+        return self.get_request(req, **kwargs)
+
+    def get_request(self, request, *, target=None, **kwargs):
+        """Start a download with a QNetworkRequest.
+
+        Args:
+            request: The QNetworkRequest to download.
+            target: Where to save the download as usertypes.DownloadTarget.
+            **kwargs: Passed to _fetch_request.
+
+        Return:
+            The created DownloadItem.
+        """
+        # WORKAROUND for Qt corrupting data loaded from cache:
+        # https://bugreports.qt.io/browse/QTBUG-42757
+        request.setAttribute(QNetworkRequest.CacheLoadControlAttribute,
+                             QNetworkRequest.AlwaysNetwork)
+
+        if request.url().scheme().lower() != 'data':
+            suggested_fn = urlutils.filename_from_url(request.url())
+        else:
+            # We might be downloading a binary blob embedded on a page or even
+            # generated dynamically via javascript. We try to figure out a more
+            # sensible name than the base64 content of the data.
+            origin = request.originatingObject()
+            try:
+                origin_url = origin.url()
+            except AttributeError:
+                # Raised either if origin is None or some object that doesn't
+                # have its own url. We're probably fine with a default fallback
+                # then.
+                suggested_fn = 'binary blob'
+            else:
+                # Use the originating URL as a base for the filename (works
+                # e.g. for pdf.js).
+                suggested_fn = urlutils.filename_from_url(origin_url)
+
+        if suggested_fn is None:
+            suggested_fn = 'qutebrowser-download'
+
+        return self._fetch_request(request,
+                                   target=target,
+                                   suggested_filename=suggested_fn,
+                                   **kwargs)
+
+    def _fetch_request(self, request, **kwargs):
+        """Download a QNetworkRequest to disk.
+
+        Args:
+            request: The QNetworkRequest to download.
+            **kwargs: passed to fetch().
+
+        Return:
+            The created DownloadItem.
+        """
+        reply = self._networkmanager.get(request)
+        return self.fetch(reply, **kwargs)
+
+    @pyqtSlot('QNetworkReply')
+    def fetch(self, reply, *, target=None, auto_remove=False,
+              suggested_filename=None, prompt_download_directory=None):
+        """Download a QNetworkReply to disk.
+
+        Args:
+            reply: The QNetworkReply to download.
+            target: Where to save the download as usertypes.DownloadTarget.
+            auto_remove: Whether to remove the download even if
+                         ui -> remove-finished-downloads is set to -1.
+
+        Return:
+            The created DownloadItem.
+        """
+        if not suggested_filename:
+            if isinstance(target, usertypes.FileDownloadTarget):
```

Comment:
All these `isinstance`s make me think there should be a `DownloadTarget.suggested_filename()` that is overridden in subclasses (and you fall back to `parse_content_disposition` if it throws a `NotImplementedError`).

There's a similar case is `isinstance`s earlier, but it would be a little more difficult there as it depends on data from the enclosing scope.